### PR TITLE
(feat) Form app: add translations for form-engine strings

### PR DESCRIPTION
**Pull request by icrc-jofrancisco**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary

This pull request adds support for translating error messages and labels within a form. Labels are now provided by a JSON file, following the naming convention ".json". This update provides a more user-friendly experience for non-native speakers.

Currently, the form supports English and French languages, and in the future, additional languages can be added by creating a new JSON file with the appropriate naming convention.

This Pull Request should be reviewed after it has been approved by the [openmrs-ngx-formentry module](https://github.com/openmrs/openmrs-ngx-formentry/pull/42).
Both Pull Requests have been tested simultaneously to ensure that both work.

## Screenshots

Here is a simple example of the translated error messages embedded in the form-entry-app
image

Translated strings for the form entry app

https://github.com/openmrs/openmrs-esm-patient-chart/assets/51502471/7f955341-c15d-4257-bfc1-dce6374cd9c9

### CODE REVIEW

File: packages/esm-form-entry-app/src/app/loaders/json-loader.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { TranslateLoader } from '@ngx-translate/core';
+import { Observable, of } from 'rxjs';
+
+export class JsonLoader implements TranslateLoader {
+  constructor() {}
+  getTranslation(lang: string): Observable<any> {
+    return of(require(`../../../translations/${lang}.json`));
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/loaders/json-loader.ts:7]:
```js
require.context(
  "../translations",
  false,
  /.json$/,
  "lazy"
);
```

Comment (ibacher) [reply packages/esm-form-entry-app/src/app/loaders/json-loader.ts:7]: Actually, better suggestion for this is probably:

```suggestion
    return of(import(`../../../translations/${lang}.json`));
```

**Review by ibacher — COMMENTED** [review]
Nit: Translation JSON's should have an empty new line at the EOF. Also, switch out `require()` for `import()` (they're both going to get re-written into a Webpack internal; we may as well use the valid ES2015 syntax).

File: packages/esm-form-entry-app/src/app/loaders/json-loader.ts (hunk 1 above)
Comment (vasharma05) [reply packages/esm-form-entry-app/src/app/loaders/json-loader.ts:7]: Hi @ibacher!
I tried with the `import`, but the application didn't pick up the translations. 🤔

**Comment by vasharma05** [comment]
Please re-review the PR. 
Thanks!

**Comment by vasharma05** [comment]
Hi @ibacher @denniskigen , please re-review.
Thanks!

File: package.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
   },
   "devDependencies": {
     "@openmrs/esm-framework": "next",
+    "@openmrs/ngx-formentry": "next",
```
Comment (ibacher) [inline package.json:20]: This should not be a dev dependency of the workspace, just the module that needs it.

File: packages/esm-form-entry-app/src/app/fe-wrapper/fe-wrapper.component.ts [hunk 3]
```diff
@@ -231,8 +237,8 @@ export class FeWrapperComponent implements OnInit, OnDestroy {
             showToast({
               critical: true,
               kind: 'success',
-              description: `The form has been submitted successfully.`,
-              title: this.form.schema.name,
+              description: this.translateService.instant('formSubmittedSuccessfully'),
+              title: this.form.schema.display,
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/fe-wrapper/fe-wrapper.component.ts:241]: There may be something we need to do here, but I don't think this is it. Specifically, `this.form.schema` refers to the schema itself, which does not have a display property. At least `name` will be defined.

File: packages/esm-form-entry-app/src/app/fe-wrapper/fe-wrapper.component.ts [hunk 4]
```diff
@@ -242,10 +248,11 @@ export class FeWrapperComponent implements OnInit, OnDestroy {
...
               kind: 'error',
-              description: `An error has occurred while submitting the form. Error: ${this.extractErrorMessagesFromResponse(
-                error,
-              )}.`,
-              title: this.form.schema.name,
+              description: this.translateService
+                .instant('formSubmissionFailed')
+                .replace('{error}', this.extractErrorMessagesFromResponse(error)),
+
+              title: this.form.schema.display,
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/fe-wrapper/fe-wrapper.component.ts:255]: Ditto on `form.schema.display`

File: packages/esm-form-entry-app/translations/en.json [hunk 5]
```diff
@@ -1 +1,66 @@
-{}
\ No newline at end of file
+{
+  "ADD_SECaddSecondOND": "Add a second",
```
Comment (ibacher) [inline packages/esm-form-entry-app/translations/en.json:2]: This looks like an error?

File: packages/esm-form-entry-app/translations/fr.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
+{
+  "requiredField": "Ce champs est requis!",
```
Comment (ibacher) [inline packages/esm-form-entry-app/translations/fr.json:2]: Could we ensure the ordering of keys is consistent across the three language files? Makes it a little easier to review.

File: packages/esm-form-entry-app/translations/en.json [hunk 7]
```diff
@@ -1 +1,66 @@
...
+  "saveAndCloseButton": "Save and close",
+  "selectWeeks": "Select Weeks",
+  "set": "Set",
+  "submitting": "Submitting",
+  "to": "To",
+  "tryOpeningAnotherForm": "Try opening another form",
+  "useValue": "Use Value",
+  "weeks": "Weeks",
+  "yearsAgo": " years ago"
+}
```
Comment (ibacher) [inline packages/esm-form-entry-app/translations/en.json:66]:
```suggestion
}

```

File: packages/esm-form-entry-app/translations/km.json [hunk 8]
```diff
@@ -1 +1,66 @@
...
+  "saveAndCloseButton": "Save and close",
+  "selectWeeks": "Select Weeks",
+  "set": "Set",
+  "submitting": "Submitting",
+  "to": "To",
+  "tryOpeningAnotherForm": "Try opening another form",
+  "useValue": "Use Value",
+  "weeks": "Weeks",
+  "yearsAgo": " years ago"
+}
```
Comment (ibacher) [inline packages/esm-form-entry-app/translations/km.json:66]:
```suggestion
}

```

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @icrc-jofrancisco! Just a little bit of clean-up and should be good to go.

**Review by ibacher — APPROVED** [review]

File: packages/esm-form-entry-app/translations/km.json
Comment (ibacher) [inline packages/esm-form-entry-app/translations/km.json:1]: @rbuisson At some point, we're going to need this file translated for km. I don't know who to ask about that.

**Pull request merged**